### DISCUSSION TITLE
Adafruit DC Motor (+ Stepper) FeatherWing

### DISCUSSION THREAD

**Discussion by questrov** [fjp/2wd-robot#50] [Q&A] [answered]
Next question, actually 2;-)  At this point have the components connected per the [Remo schematic](https://github.com/ros-mobile-robots/diffbot#hardware-architecture-and-wiring) with diffbot packages installed on Pi4.

1.) After watching the video https://ros-mobile-robots.com/teensy-mcu/#base-controller we have PlatformIO open as shown and can easily upload base controller main.cpp.  How do I now select any one of the other 3 .cpp files (test_adafruit_motoshield, test_encoders and test_adafruit_motor_controller) to build/upload to Teensy?  Would be nice to upload and test Adafruit DC Motor (+ Stepper) FeatherWing without changing any wiring except move the USB cable from SBC to Work Station.

2.)  Once Adafruit DC Motor (+ Stepper) FeatherWing is tested (it already is but standalone via Arduino Uno) how to tell ROS packages the hardware is Remo opposed to diffbot?  Tried appending the launch commands with `model:=remo` (as suggested at [Remo github](https://github.com/ros-mobile-robots/remo_description)) but still don't seem to get any output from FeatherWing.  ROS `rostopic echo /cmd_vel` shows the commands from [rqt_robot_steering](http://wiki.ros.org/rqt_robot_steering) OK.

I have searched the docs (web too) and looked thru the files but afraid to make any 'experimental' changes as that typically ends with me starting over from scratch and asking for help anyway:-/

**Comment by fjp** [comment, marked as answer]
1. You can make use of [PlatformIOs Unit Testing feature](https://docs.platformio.org/en/latest/plus/unit-testing.html). To run all the test inside the test folder you can use the "Test" task, which is found under Advanced:

![image](https://user-images.githubusercontent.com/1286618/149677050-2502bce2-a1a9-4afb-a15c-0f62dd06d007.png)

I just fixed two of the tests (`test_encoders` and `test_adafruit_motor_controller`) so please pull those latest changes. 

The first test was working already. For me the wheels are spinning during the test. It is important that you connect power to the RPi as it will provide 3.3V to the i2c connection of the Adafruit motor shield. You can probably also provide power from the MCU (Arduino or Teensy, ...) but I haven't tested that. Anyway you do it you make sure this is not an issue, otherwise the communication between MCU and motor driver won't work. The following figure shows the i2c hub in blue which is connected (indirectly through the bread board) to the 3.3V of the RPi:

![fritzing remo](https://raw.githubusercontent.com/ros-mobile-robots/ros-mobile-robots.github.io/main/docs/fritzing/remo_architecture.svg)

2. Which launch files are you using? It is also important that you have your ROS network setup correctly, explained [here](https://ros-mobile-robots.com/ros-network-setup/). After that you should be able to startup Remo robot with the following launch command:

```console
roslaunch diffbot_bringup bringup_with_laser.launch model:=remo
```

This will bring up all the hardware including the laser. Don't worry about the IMU error message, it is not yet implemented but the main algorithms will work anyway using only the encoders for odometry/localization.

Then, on your work PC you can launch SLAM using:

```console
roslaunch diffbot_slam diffbot_slam.launch
```

For this command no `model:=remo` parameter is needed. You could optionally append `slam_method:=gmapping`.

Finally, to steer the robot around for mapping your environment, you can use the keyboard teleop bringup launch file on your work PC:

```console
roslaunch diffbot_bringup keyboard_teleop.launch
```

**Reply by questrov** [reply]
Yay that all worked I have motors now!  My bad at some point pulled the 3.3V supply to FeatherWing and didn't notice in my rats nest of test wiring:-/  Thinks it time to get this mess off the test bench and build a real robot:-)

Thanks again for the help so far, much appreciated!!!
